### PR TITLE
Update rustls to 0.23

### PR DESCRIPTION
**Pull request by Ralith**
Required for Quinn to update past rustls 0.21, so a release soon after would be appreciated. Also relaxes default feature requirements on rustls.

### CODE REVIEW

**Comment by djc** [comment]
I think I tried this and it won't work on macOS...

**Comment by Ralith** [comment]
This is difficult to test without CI working by default.

**Comment by complexspaces** [comment]
@Ralith I added you with the `Triage` role on the repository (it grants minimal issue and PR permissions). I can remove it once you've merged this and GitHub doesn't see you as an "untrusted contributor" here anymore but hopefully it doesn't bug you for now.

**Comment by Ralith** [comment]
Android CI failing:

```
> Task :rustls-platform-verifier:connectedDebugAndroidTest

[DDMLIB]: An unexpected packet was received before the handshake.
[DDMLIB]: An unexpected packet was received before the handshake.
[DDMLIB]: An unexpected packet was received before the handshake.
[DDMLIB]: An unexpected packet was received before the handshake.

> Task :rustls-platform-verifier:connectedDebugAndroidTest
Starting 3 tests on test(AVD) - 9

org.rustls.platformverifier.CertificateVerifierTests > runRealWorldTestSuite[test(AVD) - 9] FAILED 
	org.junit.ComparisonFailure: A test failed. Check the logs above for Rust panics. expected:<[success]> but was:<[test failed!]>
	at org.junit.Assert.assertEquals(Assert.java:115)
Tests on test(AVD) - 9 failed: There was 1 failure(s).
```
Not sure how to interpret that. There aren't any panics that I can see. "An unexpected packet was received" sounds like a rustls-side issue. There seems to be some sort of test result artifact that might have details, but it doesn't look human-readable.

**Comment by zh-jq-b** [comment]
It seems to be the same issue as reported here https://github.com/rustls/rustls-platform-verifier/issues/68.

**Comment by cpu** [comment]
> Not sure how to interpret that. There aren't any panics that I can see. "An unexpected packet was received" sounds like a rustls-side issue. 

I think these are unrelated noise. The same output has been happening [on passing builds in the past](https://github.com/rustls/rustls-platform-verifier/actions/runs/8053565882/job/21996258520). "ddmlib" looks like it's specific to the Android emulator<->Host debug bridge and so I think likely to be inconsequential to the Rustls tests.

> It seems to be the same issue as reported here https://github.com/rustls/rustls-platform-verifier/issues/68.

Yes, it is.

File: rustls-platform-verifier/src/verification/apple.rs
Comment (Ralith) [inline rustls-platform-verifier/src/verification/apple.rs:1]: Note changes here in particular. Not sure why this was hardcoding *ring* in the past.

**Comment by Ralith** [comment]
If the only CI failure is #68, can we consider this unblocked?

File: rustls-platform-verifier/src/verification/apple.rs
Comment (cpu) [reply rustls-platform-verifier/src/verification/apple.rs:1]: Yup, exactly :+1:

**Comment by cpu** [comment]
> If the only CI failure is https://github.com/rustls/rustls-platform-verifier/issues/68, can we consider this unblocked?

I don't know how @complexspaces feels but I think we wouldn't want to cut a release until `main` is passing CI so even if we merged this it would be blocked on #68. 

I've made some progress and think I have a fix. Will update #68 shortly

**Comment by cpu** [comment]
> I've made some progress and think I have a fix. Will update https://github.com/rustls/rustls-platform-verifier/issues/68 shortly

I think cherry-picking in 98725337daa56ebe98db7877670b3886884539e8 from https://github.com/rustls/rustls-platform-verifier/pull/71 will sort the mac task out.

**Comment by Ralith** [comment]
macos is hitting:

> rustls default CryptoProvider not set

which is from the `expect` I added on [`CryptoProvider::get_default()`](https://docs.rs/rustls/latest/rustls/crypto/struct.CryptoProvider.html#method.get_default) calls. This stopped happening on other platforms when the rustls `ring` feature was enabled. I'd expect to have to call `install_default` either everywhere or nowhere; having that requirement vary across platforms is baffling.

**Comment by complexspaces** [comment]
That is extraordinarily strange. I tried comparing the output of `cargo tree --edges features` between macOS and Linux and nothing jumps out as "broken" in `rustls` direct dependency tree. The only difference is that there's no `rustls-webpki`.

**Comment by Ralith** [comment]
Perhaps it's a test case ordering/global state issue. rustls tests suggest that `None` is the expected return value even when the feature is enabled:

https://github.com/rustls/rustls/blob/0398ac50fe18521da038aa39b55eec6c83450523/rustls/tests/process_provider.rs#L47-L58

I'll try setting defaults explicitly.

**Comment by complexspaces** [comment]
It looks like its back to the Android failures 🎉 If you rebase on `main` Ci should be green at least. I merged #71 a few minutes ago.

File: rustls-platform-verifier/src/verification/windows.rs
Comment (djc) [inline rustls-platform-verifier/src/verification/windows.rs:444]: All these constructors should probably document that they're relying on the process-global crypto provider set via rustls?

File: rustls-platform-verifier/src/tests/mod.rs
Comment (djc) [inline rustls-platform-verifier/src/tests/mod.rs:65]: I think the `cfg(test)` guard can be outside the function?

**Review by cpu — COMMENTED** [review]
Reviewing the diff I don't have any feedback above what djc has already flagged. Thanks again for picking this up Ralith :tada:

File: rustls-platform-verifier/src/tests/mod.rs
Comment (complexspaces) [reply rustls-platform-verifier/src/tests/mod.rs:65]: ++ on this. We shouldn't need the guard at all because the `tests` module should be under the same thing already.

Comment (Ralith) [reply rustls-platform-verifier/src/tests/mod.rs:65]: That's what I thought, too. However, the `tests` module is gated by `#[cfg(any(test, feature = "ffi-testing"))]`, and whatever builds it with `ffi-testing` in CI isn't a test build, so it doesn't get the *ring* dev-dependency, so the build fails. I think this is also related to the remaining failure, in which the android tests presumably don't set up a default provider. Haven't actually found the android tests or worked out a strategy to address that yet. Maybe we need some sort of automatic fallback like `rustls` seems to have.

Comment (complexspaces) [reply rustls-platform-verifier/src/tests/mod.rs:65]: That looks like it it actually, good catch. With the improved Android logging I saw it was throwing the missing crypto provider error for every test. Locally I updated the `ffi-testing` feature and made sure to call `ensure_global_state` in the Android tests. After that everything worked. Here's the changes if you'd like to commit them:
```diff
diff --git a/rustls-platform-verifier/Cargo.toml b/rustls-platform-verifier/Cargo.toml
index cdffcdf..dc2f369 100644
--- a/rustls-platform-verifier/Cargo.toml
+++ b/rustls-platform-verifier/Cargo.toml
@@ -19,7 +19,7 @@ crate-type = ["cdylib", "rlib"]
 # Enables a C interface to use for testing where `cargo` can't be used.
 # This feature is not stable, nor is the interface exported when it is enabled.
 # Do not rely on this or use it in production.
-ffi-testing = ["android_logger"]
+ffi-testing = ["android_logger", "rustls/ring"]
 # Enables APIs that expose lower-level verifier types for debugging purposes.
 dbg = []
 # Enables `log::debug` base64-encoded logging of all end-entity certificates processed
diff --git a/rustls-platform-verifier/src/tests/ffi.rs b/rustls-platform-verifier/src/tests/ffi.rs
index bf18a01..6a4a5af 100644
--- a/rustls-platform-verifier/src/tests/ffi.rs
+++ b/rustls-platform-verifier/src/tests/ffi.rs
@@ -47,6 +47,7 @@ mod android {
                     .with_filter(log_filter),
             );
             crate::android::init_hosted(env, cx).unwrap();
+            crate::tests::ensure_global_state();
             std::panic::set_hook(Box::new(|info| {
                 let msg = if let Some(msg) = info.payload().downcast_ref::<&'static str>() {
                     msg
diff --git a/rustls-platform-verifier/src/tests/mod.rs b/rustls-platform-verifier/src/tests/mod.rs
index 63e6490..16ff02a 100644
--- a/rustls-platform-verifier/src/tests/mod.rs
+++ b/rustls-platform-verifier/src/tests/mod.rs
@@ -62,9 +62,6 @@ pub(crate) fn verification_time() -> pki_types::UnixTime {
     pki_types::UnixTime::since_unix_epoch(Duration::from_secs(1_710_189_025))
 }
 
-fn ensure_global_state() {
-    #[cfg(test)]
-    {
-        _ = rustls::crypto::ring::default_provider().install_default();
-    }
+pub(crate) fn ensure_global_state() {
+    let _ = rustls::crypto::ring::default_provider().install_default();
 }
```

File: rustls-platform-verifier/src/verification/windows.rs
Comment (Ralith) [reply rustls-platform-verifier/src/verification/windows.rs:444]: I think proceeding with this design would be user-hostile, as a typical use is likely to panic at runtime for users who haven't explicitly installed a default, which isn't normally required by rustls. I think a better solution would be for rustls to expose `CryptoProvider::get_default_or_install_from_crate_features`, which would let us Just Work as well as rustls itself does (although I have mixed feelings about how that breaks if multiple providers are enabled). 

We could also reimplement the implicit default logic with local features in this crate, but that would put users at higher risk of accidentally enabling multiple providers (and potentially breaking any part of their code that relies on implicit defaults) if they don't take care to match features on rustls and this crate.

Comment (djc) [reply rustls-platform-verifier/src/verification/windows.rs:444]: I think this crate should probably just take an Arc<CryptoProvider> as an argument.

Comment (Ralith) [reply rustls-platform-verifier/src/verification/windows.rs:444]: That works, though it undermines the convenience of rustls's implicit defaults if an idiomatic user has to go and deal with explicit provider handling for the verifier anyway.

Comment (complexspaces) [reply rustls-platform-verifier/src/verification/windows.rs:444]: > My other thought on this is: perhaps we should aim for windows.rs to use CryptVerifyMessageSignature (or similar) and for apple.rs to use SecKeyVerifySignature?

I have thought about this before, but I'm still not sure I like the idea. It adds another "split" in the library's functionality set, especially on Windows where registry controls can influence what cryptographic operations the OS prefers/will use. This would make debugging harder, for example. So far I have only seen one cryptographic issue come in via bug report at work. Everything else has been missing trust roots, malformed X.509, etc.

Comment (Ralith) [reply rustls-platform-verifier/src/verification/windows.rs:444]: > It adds another "split" in the library's functionality set

Isn't exposing platform-specific behavior the whole point of this lib?

> I think it might be cleaner if this operates at the level of a "parent" Arc<dyn ServerCertVerifier> rather than a CryptoProvider.

This makes a lot of sense to me.

> But unfortunately right now WebPkiServerVerifier::builder() does require a non-empty set of roots, which makes it annoying to reuse for this purpose.

Could we relax this, or introduce an alternative verifier in rustls that has an empty root set but implements the other methods? Practically speaking this amounts to creatively delegating to rustls's `get_default_or_install_from_crate_features`, which I think is the best user experience.

**Comment by Ralith** [comment]
I've added a commit that makes calls to `CryptoProvider::get_default()` lazy, which should ensure a good user experience without hard-coding a specific provider or breaking the API.

Long-term, I think it would be better for rustls to expose `CryptoProvider::get_default_or_install_from_crate_features` directly, since its use in e.g. `WebPkiServerVerifier` seems to be a tacit admission that it's the most user-friendly way for library code to get a provider. However, this change will let rustls-platform-verifier update and unblock downstream users like Quinn immediately without regressing usability.

File: rustls-platform-verifier/Cargo.toml
Comment (cpu) [inline rustls-platform-verifier/Cargo.toml:32]: Was it intentional to drop the `tls12` feature activation?

Comment (cpu) [inline rustls-platform-verifier/Cargo.toml:32]: Was it intentional to drop the `tls12` feature activation?

**Review by cpu — APPROVED** [review]

File: rustls-platform-verifier/Cargo.toml
Comment (Ralith) [reply rustls-platform-verifier/Cargo.toml:32]: Yep, very intentional. For example, Quinn users will typically not want TLS 1.2 support built. It didn't actually seem to be needed, anyway.

Comment (cpu) [reply rustls-platform-verifier/Cargo.toml:32]: Makes sense, just wanted to check explicitly 👍 Thanks!

**Comment by Lewiscowles1986** [comment]
Please merge

**Comment by ibigbug** [comment]
are we good to go with this? this seems to be the only blocker of https://github.com/quinn-rs/quinn/pull/1715 ?

**Review by complexspaces — APPROVED** [review]
Sorry for the delay, LGTM.

**Pull request merged**
